broker-daemon/block-order-worker/index.spec.js

```diff
@@ -2115,22 +2118,94 @@ describe('BlockOrderWorker', () => {
+      let workBlockOrderStub
       let failBlockOrderStub
+      let relayerIsAvailableStub
+      let getBlockOrderStub
+      let updatedBlockOrder
       let rejectListener
+      let retry
 
       beforeEach(() => {
+        workBlockOrderStub = sinon.stub().resolves()
+        worker.workBlockOrder = workBlockOrderStub
+
         failBlockOrderStub = sinon.stub().resolves(true)
         worker.failBlockOrder = failBlockOrderStub
+
+        relayerIsAvailableStub = sinon.stub().resolves(true)
+        worker.relayerIsAvailable = relayerIsAvailableStub
+
+        updatedBlockOrder = blockOrder
+        getBlockOrderStub = sinon.stub().resolves(updatedBlockOrder)
+        worker.getBlockOrder = getBlockOrderStub
+
+        retry = sinon.stub()
+        BlockOrderWorker.__set__('retry', retry)
+
         rejectListener = onceStub.withArgs('reject').args[0][1]
       })
 
       it('registers an event on an order for order rejection', async () => {
         expect(onceStub).to.have.been.calledWith('reject', sinon.match.func)
       })
 
-      it('fails a block order if the call is rejected', async () => {
-        await rejectListener()
-        expect(failBlockOrderStub).to.have.been.calledOnce()
+      context('should retry', () => {
+        beforeEach(() => {
+          order.shouldRetry.returns(true)
+        })
+
+        it('retries the block order after the relayer has gone down and becomes available', async () => {
+          await rejectListener()
+          expect(retry).to.have.been.calledOnce()
+          expect(retry).to.have.been.calledWith(sinon.match.func, sinon.match.string, sinon.match.number, sinon.match.number)
+        })
+
+        it('works the correct block order when it should retry', async () => {
+          relayerIsAvailableStub.resolves(true)
+
+          await rejectListener()
+
+          const retryBlockOrderStub = retry.args[0][0]
+          await retryBlockOrderStub()
+
+          expect(getBlockOrderStub).to.have.been.calledOnce()
+          expect(getBlockOrderStub).to.have.been.calledWith(blockOrder.id)
+
+          expect(workBlockOrderStub).to.have.been.calledOnce()
+          expect(workBlockOrderStub).to.have.been.calledWith(updatedBlockOrder, Big(order.order.baseAmount))
+        })
+
+        it('fails the block order when the retry logic throws', async () => {
+          order.order.error = 'an error'
+          retry.throws()
+          await rejectListener()
+
+          expect(failBlockOrderStub).to.have.been.calledOnce()
+          expect(failBlockOrderStub).to.have.been.calledWith(blockOrder.id, 'an error')
+        })
+
+        it('does not try re-placing the block order when relayer is not available', async () => {
+          relayerIsAvailableStub.resolves(false)
+
+          await rejectListener()
+          const retryStub = retry.args[0][0]
+          expect(retryStub()).to.have.been.rejectedWith('not available')
+        })
+      })
+
+      context('should NOT retry', () => {
```

Comment:
make sure to have a test showing that the block order was NOT re-worked in this case